src/Features/Core/Portable/InlineHints/AbstractInlineParameterNameHintsService.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable disable
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.InlineHints
+{
+    internal abstract class AbstractInlineParameterNameHintsService : IInlineParameterNameHintsService
+    {
+        public async Task<ImmutableArray<InlineParameterHint>> GetInlineParameterNameHintsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+
+            var forParameters = options.GetOption(InlineHintsOptions.EnabledForParameters);
+            if (!forParameters)
+                return ImmutableArray<InlineParameterHint>.Empty;
+
+            var literalParameters = options.GetOption(InlineHintsOptions.ForLiteralParameters);
+            var objectCreationParameters = options.GetOption(InlineHintsOptions.ForObjectCreationParameters);
+            var otherParameters = options.GetOption(InlineHintsOptions.ForOtherParameters);
+            if (!literalParameters && !objectCreationParameters && !otherParameters)
+                return ImmutableArray<InlineParameterHint>.Empty;
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var nodes = root.DescendantNodes(textSpan);
+
+            using var _1 = ArrayBuilder<InlineParameterHint>.GetInstance(out var buffer);
+            AddAllParameterNameHintLocations(semanticModel, nodes, buffer, cancellationToken);
```

Comment:
Rather than allocating a large array with everything (potentially twice), what if we give AddParameterNameHintLocations a lambda that it calls back to add something, and that lambda does the filter? The lamdba allocation would sure be cheaper than two array builders and the underlying arrays.

Reply:
yup.  that makes sense.